src/main/kotlin/exercise3/task4/TaxiParkTask.kt

```diff
@@ -10,29 +10,49 @@ package exercise3.task4
  * Find all the drivers who performed no trips.
  */
 internal fun TaxiPark.findFakeDrivers(): Set<Driver> {
-    TODO("Implement me!!!")
+    return this.allDrivers.minus(this.trips.map { it.driver }
+        .intersect(this.allDrivers))
 }
 
 /**
  * Subtask 2:
  * Find all the clients who completed at least the given number of trips.
  */
 internal fun TaxiPark.findFaithfulPassengers(minTrips: Int): Set<Passenger> {
-    TODO("Implement me!!!")
+    println(this.trips.flatMap { it.passengers }
+        .groupingBy { it }
+        .eachCount())
+
+    return this.trips.flatMap { it.passengers }
+        .groupingBy { it }
+        .eachCount()
+        .filter { it.value >= minTrips }
+        .map { it.key }
+        .toSet()
 }
 
 /**
  * Subtask 3:
  * Find all the passengers, who were taken by a given driver more than once.
  */
 internal fun TaxiPark.findFrequentPassengers(driver: Driver): Set<Passenger> {
-    TODO("Implement me!!!")
+    return this.trips
+        .filter { it.driver == driver }
+        .flatMap { it.passengers }
+        .groupingBy { it }
+        .eachCount()
+        .filter { it.value > 1 }
+        .map { it.key }
+        .toSet()
 }
 
 /**
  * Subtask 4:
  * Find the passengers who had a discount for the majority of their trips.
  */
 internal fun TaxiPark.findSmartPassengers(): Set<Passenger> {
-    TODO("Implement me!!!")
+    return this.trips
+        .filter { it.discount != null }
+        .flatMap { it.passengers }
```

Comment:
Your solution returns a set of passengers who had at least one trip with the discount.
The assignment request is to return a list of passengers who had more than half of the trips with the discount.